docs/rfcs/016-user-and-database-management.md

```diff
@@ -0,0 +1,70 @@
+## Zentih user and database management RFC
+
+### CREATE USER
+
+CREATE USER is only allowed from console, 
+because console is involved in the auth process, when users connect via proxy.
```

Comment:
I think our proxy should not remain the main point of authentication for clusters. Once a cluster has started, the connections should go directly to the cluster whenever possible to reduce latency and overhead.

Console should not be involved in the authentication process. There is 0 reason for that to happen; it only complicates things: Why would we even _want_ to do a bespoke user authentication for PostgreSQL -- it only adds specific behaviors that clients would need to be able to handle and that we would need to document and maintain.

It would be amazing to keep Proxy as _only_ a transparent proxy for those connections that support it, and it would save a lot of work on all sides.

Reply:
> Console should not be involved in the authentication process. There is 0 reason for that to happen

There is at least 1 reason for that: if we want 'serverless', i.e. spin-up computes when they are needed and tear down them when they are not, and 'needed' means someone wants to connect. Then we have to accept this connection attempts, wake up compute and route traffic to it

Reply:
I do not consider "waking up compute nodes" part of the authentication process but part of setting up the connection (a small but significant difference). Console does not need to know what authentication details are being used to connect to compute node Y; it only needs to know that compute node Y needs to wake up.

Reply:
> Console does not need to know what authentication details are being used to connect to compute node Y; it only needs to know that compute node Y needs to wake up.

Then I can spam connection attempts to your cluster, prevent it from being suspended and burn your money. THat's just an example.

And I'm not sure about technical part as well. You propose to don't do auth on proxy or console, then it won't behave like MITM. But when you accept a libpq connection, you basically start the auth flow, so I'm not sure you can say 'wait a second, I've just woken up your Postgres, can you please start all the same auth sequence, but with another server?'

Or you propose that our proxy shouldn't be a libpq proxy? TCP?

Reply:
> Then I can spam connection attempts to your cluster, prevent it from being suspended and burn your money.

Assuming there is a public endpoint to your database and your database access rules allow access to your cluster, then yes.

> Or you propose that our proxy shouldn't be a libpq proxy? TCP?

Yep, basically not much more than a libpq-aware TCP or TLS proxy. I.e., when we implement SNI like "psql -h cluster-name-123456.dbs.zenith.tech", it would just wake up cluster-name-123456 and forward the connection, IP ranges permitting.

Maybe we could add some extra goodies for authentication, such as not waking up the whole PG cluster but only the authentication part, but that's the extent of what I think proxy needs. And for those authentication goodies, there's no real need for Console to authenticate the client or maintain and synchronize a database of users for PostgreSQL Cluster Y.

Reply:
> Maybe we could add some extra goodies for authentication, such as not waking up the whole PG cluster but only the authentication part, but that's the extent of what I think proxy needs.

And exactly for that, you need to decipher traffic and have auth info somewhere not in the destination postgres. Isn't it?

Reply:
... only if the target cluster is currently offline, in which case you'd already need to contact our control plane. But in the happy case (cluster is running), we shouldn't need to contact Console IMO.

Reply:
> ... only if the target cluster is currently offline, in which case you'd already need to contact our control plane. But in the happy case (cluster is running), we shouldn't need to contact Console IMO.

And this means that if the cluster is ever can be offline we need to always have a synchronous copy of roles info in the console. If the cluster is online we can go directly to the postgres, but that is a tradeoff between 'guard postgres from exhausting max_connection with bruteforce' vs 'have a smaller connection latency'.

Reply:
It could be, but we can minimize any possible downtime later by separating HTTP API service from the current monolithic console. After that it will be a stateless web service, which should be rather durable and less error-prone considering all the k8s automation for high availability